model/task_queue_service_dependency.go

```diff
@@ -457,15 +436,59 @@ func (d *basicCachedDAGDispatcherImpl) FindNextTask(ctx context.Context, spec Ta
 	return nil
 }
 
+// tryMarkItemDispatched will dispatch a standalone task if all of the following are true:
+// (a) it's not marked as dispatched in the in-memory queue.
+// (b) a record of the task exists in the database.
+// (c) it never previously ran on another host.
+// (d) all of its dependencies are satisfied.
+// Returns true if the item has already been marked dispatched by a separate request.
+func (d *basicCachedDAGDispatcherImpl) tryMarkItemDispatched(item *TaskQueueItem) bool {
+	d.mu.Lock()
+	defer d.mu.Unlock()
+
+	if item.IsDispatched {
+		return true
+	}
+
+	// Cache the task as dispatched from the in-memory queue's point of view.
+	// However, it won't actually be dispatched to a host if it doesn't satisfy all constraints.
+	item.IsDispatched = true // *TaskQueueItem
+	return false
+}
+
+func (d *basicCachedDAGDispatcherImpl) getNextTaskForTaskGroup(taskGroupUnit schedulableUnit) *TaskQueueItem {
```

Comment:
nit: consider calling `tryMarkNextTaskGroupTaskDispatched` or something like that, since it's quite similar to `tryMarkItemDispatched` but for task group tasks.